File: pocketcloud/poc/src/crypto/beaver.js
// Beaver multiplication triples.
//
// Secret x secret multiplication needs one precomputed triple (a, b, c)
// with c = a*b, dealt as MAC'd sharings. Online phase:
//   1. workers open d = x - a and e = y - b (both uniform random, so the
//      openings leak nothing about x or y); the verifier MAC-checks the
//      openings themselves — an incorrectly opened d is equivalent to
//      silently substituting a different input and MUST be caught here;
//   2. each worker computes z_i = c_i + d*b_i + e*a_i, and the same
//      combination on the MAC shares;
//   3. the verifier adds the public terms d*e (to z) and alpha*d*e (to the
//      MAC) at reconstruction, then checks mac == alpha * z.
//
// In this PoC the coordinator is the trusted dealer (the PRD's "managed
// mode", Phase 1). Distributed triple generation removes that trust later.

import { mul, randomFieldElement } from './field.js';
import { dealWithMac } from './secret-sharing.js';

// Generate `count` triples dealt to n workers under MAC key alpha.
// Returns perWorker[i] = { a: {share, mac}, b: {...}, c: {...} } where each
// share/mac is a vector of length `count` (one slot per multiplication).
export function dealTriples(count, n, alpha) {
  const a = Array.from({ length: count }, () => randomFieldElement());
  const b = Array.from({ length: count }, () => randomFieldElement());
  const c = a.map((ak, k) => mul(ak, b[k]));

  const aDealt = dealWithMac(a, n, alpha);
  const bDealt = dealWithMac(b, n, alpha);
  const cDealt = dealWithMac(c, n, alpha);

  return Array.from({ length: n }, (_, i) => ({
    a: aDealt[i],
    b: bDealt[i],
    c: cDealt[i],
  }));
}

File: pocketcloud/crypto/src/beaver.rs
//! Beaver multiplication triples (SPEC-001 §5).
//!
//! Secret x secret multiplication needs one precomputed triple (a, b, c=ab),
//! dealt as MAC'd sharings. v0 uses a trusted dealer (managed mode); distributed
//! generation is CC-060. The online phase (open d=x-a, e=y-b with MAC checks on
//! the openings, then combine) lives in the coordinator/verifier — this module
//! deals the triples.

use crate::field::Field;
use crate::rng::FieldRng;
use crate::sharing::{deal_with_mac, MacShare};

/// A worker's slice of `count` triples: shares of a, b, and c = a*b.
pub struct TripleShare<F: Field> {
    /// Shares (value+MAC) of the `a` vector.
    pub a: MacShare<F>,
    /// Shares (value+MAC) of the `b` vector.
    pub b: MacShare<F>,
    /// Shares (value+MAC) of the `c = a*b` vector.
    pub c: MacShare<F>,
}

/// Deal `count` triples to `n` workers under MAC key `alpha`.
pub fn deal_triples<F: Field, R: FieldRng>(
    count: usize,
    n: usize,
    alpha: F,
    rng: &mut R,
) -> Vec<TripleShare<F>> {
    let a: Vec<F> = (0..count).map(|_| rng.next_field::<F>()).collect();
    let b: Vec<F> = (0..count).map(|_| rng.next_field::<F>()).collect();
    let c: Vec<F> = a.iter().zip(&b).map(|(&ak, &bk)| ak.mul(bk)).collect();

    let a_dealt = deal_with_mac(&a, n, alpha, rng);
    let b_dealt = deal_with_mac(&b, n, alpha, rng);
    let c_dealt = deal_with_mac(&c, n, alpha, rng);

    a_dealt
        .into_iter()
        .zip(b_dealt)
        .zip(c_dealt)
        .map(|((a, b), c)| TripleShare { a, b, c })
        .collect()
}
